[lldb][NFC] Make the target's SectionLoadList private.

clayborg opened the pull request:
Lots of code around LLDB was directly accessing the target's section load list. This NFC patch makes the section load list private so the Target class can access it, but everyone else now uses accessor functions. This allows us to control the resolving of addresses and will allow for functionality in LLDB which can lazily resolve addresses in JIT plug-ins with a future patch.

clayborg committed:
Make the target's SectionLoadList private.

Lots of code around LLDB was directly accessing the target's section load list. This NFC patch makes the section load list private to the Target class can access it, but everyone else now uses accessor functions. This allows us to control the resolving of addresses and will allow for functionality in LLDB which can lazily resolve addresses in JIT plug-ins with a future patch.

JDevlieghere commented on lldb/source/Target/Target.cpp:
```diff
@@ -5066,3 +5066,17 @@ llvm::json::Value
 }
+
+bool Target::SectionLoadListIsEmpty() const {
+  return const_cast<Target *>(this)->GetSectionLoadList().IsEmpty();
```
To avoid the `const_cast`, can we have a `const` `GetSectionLoadList` overload? `SectionLoadHistory::IsEmpty()` is already `const`.

clayborg replied:
We actually can't because `SectionLoadHistory::GetCurrentSectionLoadList()` needs to be able to create an empty section load for the special stop ID of `eStopIDNow` if the section load list is empty. I don't thing we want to change section load list to have the load list be mutable. Let me know if you disagree or have another solution?

clayborg replied:
I found a way to fix this. We don't need the  this function to be `const`

JDevlieghere reviewed (commented):
Protecting the section load list seems reasonable. I left one inline comment + this needs to be `clang-format`'ed.

jeffreytan81 reviewed (approved):
LGTM. Pending Jonas's comment and clang-format

jimingham commented:
This is just naming, but it seems odd to have hidden the SectionLoadList in Target, to then have Target::SectionLoadListIsEmpty() as a primary API.  "What is this SectionLoadList thing of which you speak???"

Maybe "HasLoadedSections"?

jimingham commented:
BTW, the general change seems fine to me, none of the clients actually needed to know about the list, they were just asking "can I use load addresses" and "resolve this load address for me" so they didn't need to know how that was done.

clayborg committed:
Respond to user feedback:

- Rename Target::SectionLoadListIsEmpty() to Target::HasLoadedSections() and invert all call sites
- Change Target::HasLoadedSections() to not be const and avoid the const cast.

clayborg commented:
Ok, all comments were addressed.

jimingham commented on lldb/include/lldb/Target/Target.h:
```diff
@@ -1140,9 +1140,13 @@ class Target : public std::enable_shared_from_this<Target>,
+
+  lldb::addr_t GetSectionLoadAddress(const lldb::SectionSP &section_sp);
+
+  void ClearSectionLoadList();
```
It seems like nobody but dynamic loaders should be calling these two.  Is there some way to express that?

clayborg replied:
`GetSectionLoadAddress(...)` is called all over the place, but `ClearSectionLoadList()` is called only by `DynamicLoaderMacOS`, but it should be able to be called by any dynamic loader, so I wouldn't want to limit the ability to call it. And we can't tie it to a dynamic loader as I could see us wanting to add a command that would call this function for symbolication purposes.

jimingham replied:
Okay.

clayborg committed:
Fix issue with test suite.

Now that we are using Target::ResolveLoadAddress, we had an implicit conversion going on that I found that was causing many failures in the test suite. Fixed now.

clayborg committed:
Fixed two remaining test suite issues with this patch.
